RP2040: fix USB stability issue related to control transfers.

purdeaandrei opened the pull request:
Description of the bad behaviour before this change:

The arming of the control EP0_OUT endpoint was not kept under control. It could happen that the EP0_OUT endpoint was left armed, after the completion of a complete control transfer.

It is clear that the intention was to NOT keep EP0_OUT constantly armed while idle, because usb_dc_ep_enable() doesn't arm it, and the intention was for when usb_dc_ep_read() is called to collect the Setup-Stage 8-byte data, that is when EP0_OUT is armed, and before this call is performed, the host will keep getting NAKs for the Data-Stage of the to_device control transfer.

This happens correctly on the first to_device control transfer with wLength > 0. However, because usb_dc_ep_read_continue() indiscriminately re-arms all OUT endpoints, in the case of to_device control transfers with wLength > 0, on the Data-Stage, the endpoint is also re-armed, which is wrong, because then the endpoint will be left armed after the control transfer is over.

In this case when a new to_device control transfer starts, the Data-Stage will be accepted on the first try. This would still have worked without a failure if the Setup-Stage would have been processed immediately, but because we process everything in a work queue at a later time, when the Setup-Stage associated 8-byte data buffer is read both the Setup-Stage and Data-Stage have arrived. At the end of handling the Setup-Stage we try to re-arm the EP0_OUT, which already contains data, thereby corrupting the received length portion of the buf_ctl register. (Obviously other fields are changed too, but the length field is the one that first causes chaos, cause it's written to the maximum, which is 64.) The above mentioned Data-Stage already has a message in its workqueue for it to be processed, but it is picked up only after the length field has been corrupted. Because of this usb_dc_ep_read() thinks there is more data in the buffer than there really is, and everything becomes de-synchronized, with later reads accessing uninitialized parts of the buffer.

This sounds like a fundamental failure, that should make it impossible to operate USB, however the reason this behaviour doesn't make it impossible to enumerate the device is that this only affects to_device control transfers with wLength > 0, and during enumeration there are not many of those happening.
When enumerating a HID keyboard, there is only _one_ of those happening, and it is the initial setting of the lock light led status. And that first one succeeds because it's the first one. (However, later lock light setting control transfers can cause problems, which is how this problem was encountered.)

The solution in this commit is to keep better control over when EP0_OUT is armed. This forces the Data-Stage to arrive later (the host will keep re-trying), and that way the corruption of the buffer control register is avoided.

Summary of the changes:
- Rework the logic around deciding wether to re-arm the out endpoint after a read.
For non-0 endpoint the previous behaviour is kept, however for EP0 it is only re-
armed if more OUT transactions are expected for that SETUP transfer (be it data-
stage or status-stage)
- Force un-arm the EP0_OUT endpoint in case a stall condition is observed.
- When a setup transfer is received check if EP0_OUT is already armed. If armed
then log a warning message, and force-disarm it.
- When a setup req interrupt fires, don't immediately force the next
       read to get it, instead, it will be read only after a setup message
       is extracted from the message queue.
- When a setup packet is received abort any unfinished previous control
       transfers:
        - cancel any data buffers given to the EP0_IN endpoint
        - drop any new ep0_in writes that are attempted before this newest
          setup packet's associated message is extracted from the message
          queue.
- In the ISR, check buffer interrupts before setup req interrupts.
       This is to make sure that the final 0-length status message from the
       previous setup packet is consumed before the new setup packet.
       (this is the only case now when both interrupts could be seen as
       having fired by the time the interrupt handler routine executes.

purdeaandrei committed:
drivers: usb_dc_rpi_pico: fix stability issue related to control transfers

Description of the bad behaviour before this change:

The arming of the control EP0_OUT endpoint was not kept under
control. It could happen that the EP0_OUT endpoint was left
armed, after the completion of a complete control transfer.

It is clear that the intention was to NOT keep EP0_OUT constantly
armed while idle, because usb_dc_ep_enable() doesn't arm it,
and the intention was for when usb_dc_ep_read() is called to
collect the Setup-Stage 8-byte data, that is when EP0_OUT is armed,
and before this call is performed, the host will keep getting NAKs
for the Data-Stage of the to_device control transfer.

This happens correctly on the first to_device control transfer with
wLength > 0. However, because usb_dc_ep_read_continue() indiscriminately
re-arms all OUT endpoints, in the case of to_device control transfers
with wLength > 0, on the Data-Stage, the endpoint is also re-armed,
which is wrong, because then the endpoint will be left armed after
the control transfer is over.

In this case when a new to_device control transfer starts, the
Data-Stage will be accepted on the first try. This would still
have worked without a failure if the Setup-Stage would have been
processed immediately, but because we process everything in a work
queue at a later time, when the Setup-Stage associated 8-byte data
buffer is read both the Setup-Stage and Data-Stage have arrived.
At the end of handling the Setup-Stage we try to re-arm the EP0_OUT,
which already contains data, thereby corrupting the received length
portion of the buf_ctl register. (Obviously other fields are changed
too, but the length field is the one that first causes chaos, cause
it's written to the maximum, which is 64.) The above mentioned Data-Stage
already has a message in its workqueue for it to be processed, but
it is picked up only after the length field has been corrupted.
Because of this usb_dc_ep_read() thinks there is more data in the buffer
than there really is, and everything becomes de-synchronized, with
later reads accessing uninitialized parts of the buffer.

This sounds like a fundamental failure, that should make it impossible
to operate USB, however the reason this behaviour doesn't make it
impossible to enumerate the device is that this only affects
to_device control transfers with wLength > 0, and during enumeration
there are not many of those happening.
When enumerating a HID keyboard, there is only _one_ of those
happening, and it is the initial setting of the lock light led status.
And that first one succeeds because it's the first one. (However, later
lock light setting control transfers can cause problems, which is how
this problem was encountered.)

The solution in this commit is to keep better control over when EP0_OUT
is armed. This forces the Data-Stage to arrive later (the host will keep
re-trying), and that way the corruption of the buffer control register
is avoided.

Summary of the changes:
 - Rework the logic around deciding wether to re-arm the out endpoint
   after a read. For non-0 endpoint the previous behaviour is kept,
   however for EP0 it is only re-armed if more OUT transactions are
   expected for that SETUP transfer (be it data-stage or status-stage)
 - Force un-arm the EP0_OUT endpoint in case a stall condition is observed.
 - When a setup transfer is received check if EP0_OUT is already armed.
   If armed then log a warning message, and force-disarm it.
 - When a setup req interrupt fires, don't immediately force the next
   read to get it, instead, it will be read only after a setup message
   is extracted from the message queue.
 - When a setup packet is received abort any unfinished previous control
   transfers:
    - cancel any data buffers given to the EP0_IN endpoint
    - drop any new ep0_in writes that are attempted before this newest
      setup packet's associated message is extracted from the message
      queue.
 - In the ISR, check buffer interrupts before setup req interrupts.
   This is to make sure that the final 0-length status message from the
   previous setup packet is consumed before the new setup packet.
   (this is the only case now when both interrupts could be seen as
   having fired by the time the interrupt handler routine executes.

Signed-off-by: Purdea Andrei <andrei@purdea.ro>
Co-authored-by: Johann Fischer

yonsch commented:
Perhaps this is related to #52705

desowin commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -635,15 +635,27 @@ int usb_dc_ep_read_continue(uint8_t ep)
+	if (state.setup_available) {
+		const struct usb_setup_packet *setup = (const void *)&usb_dpram->setup_packet;
+		if ((setup->wLength != 0U) ||
+		    !usb_reqtype_is_to_host(setup)) {
```
I think it would be clearer if you used `usb_reqtype_is_to_device(setup)` instead.

purdeaandrei replied:
Will do, well, this particular line will disappear, but will double-check to not use !X() if instead the equivalent Y() is available without being negated

purdeaandrei replied:
Done, no more usage of !usb_reqtype_is_to_host(

desowin commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -635,15 +635,27 @@ int usb_dc_ep_read_continue(uint8_t ep)
+	bool data_packet_after_setup_is_expected = false;
+	if (state.setup_available) {
+		const struct usb_setup_packet *setup = (const void *)&usb_dpram->setup_packet;
+		if ((setup->wLength != 0U) ||
```
Shouldn't the condition here be `&&` instead of `||`?

When `wLength` is 0 then the status stage is always IN ZLP and direction is ignored according to Universal Serial Bus Specification Revision 2.0 9.3.1 bmRequestType
> The state of the Direction bit is ignored if the wLength field is zero, signifying there is no Data stage.

purdeaandrei replied:
Okay, I have refactored the PR a lot, I fixed the issue you mentioned in this thread, and now I'm also covering some conditions that I have not covered before. I have also done more detailed testing on this PR than before.
I have also confirmed that it has a massive positive inpact on issue #52705 (but this issue also needs another change related to data0/data1 sequencing which I will submit as a separate PR very soon).

I have also added the following to the commit message:
    Summary of the changes:
     - Rework the logic around deciding wether to re-arm the out endpoint after a read.
       For non-0 endpoint the previous behaviour is kept, however for EP0 it is only re-
       armed if more OUT transactions are expected for that SETUP transfer (be it data-
       stage or status-stage)
     - Force un-arm the EP0_OUT endpoint in case a stall condition is observed.
     - When a setup transfer is received check if EP0_OUT is already armed. If armed
       then log a warning message, and force-disarm it.

desowin commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -635,15 +635,27 @@ int usb_dc_ep_read_continue(uint8_t ep)
-	bytes_received = state.setup_available ?
-			 sizeof(struct usb_setup_packet) :
-			 udc_rpi_get_ep_buffer_len(ep);
+	bool data_packet_after_setup_is_expected = false;
```
Move the declaration above to the declarations above to pass compliance check. Maybe rename to `arm_out_endpoint` but there can be multiple disagreements about the names...

purdeaandrei replied:
Sounds good to me, will use that name, or similar.

purdeaandrei replied:
Done, using arm_out_endpoint

desowin commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -635,15 +635,27 @@ int usb_dc_ep_read_continue(uint8_t ep)
 	LOG_DBG("received %d offset: %d", bytes_received, ep_state->read_offset);
-	if (bytes_received == ep_state->read_offset) {
+
+	if (((ep != USB_CONTROL_EP_OUT) && (udc_rpi_get_ep_buffer_len(ep) == ep_state->read_offset)) || 
```
Isn't `(ep != USB_CONTROL_EP_OUT) && (udc_rpi_get_ep_buffer_len(ep) == ep_state->read_offset)` always false due to `ep` guaranteed to be `USB_CONTROL_EP_OUT` at this point? Note that the code at beginning of this function will return `-EINVAL` when `!USB_EP_DIR_IS_OUT(ep)` is true.

desowin reviewed (changes requested):
Commit description is excellent. The code would need some modifications.

purdeaandrei commented:
Note that this change also introduces a compiler warning which is solved by PR #54193
Let me know if that is a problem, I can either remove the const from this change, or add the content of PR #54193 into this PR.

purdeaandrei commented:
@desowin @yonsch @petejohanson ready for re-review. Also you may be interested in: #54193, #54192, #54194

petejohanson commented:
I can confirm this fixes some stability issues I've encountered when using CDC logging + HID on ZMK (when cherry picked into our 3.2 work).

desowin commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -625,26 +651,66 @@ int usb_dc_ep_read_wait(uint8_t ep, uint8_t *data,
+					/* If no more data in the buffer, but we're still waiting
+					 * for more, start a new read transaction.
+					 */
+					if (bytes_received == ep_state->read_offset) {
```
I don't understand this line. When can it be not equal? Is this just checking for a short data packet (i.e. not bMaxPacketSize0)?

purdeaandrei replied:
This allows us to receive for example a single packet of 64 bytes, but then proceed to read it out in multiple pieces by calling usb_dc_ep_read() repeatedly with smaller max_data_len. When there is no data remaining in the 64 byte buffer, only then can we re-arm the endpoint, otherwise a new packet coming in could overwrite it. The only exception to this behavior is when reading in the 8 bytes of data associated with a setup transaction of a control transfer, in which case any bytes left over will be lost, all 8 bytes must be read at once if you need them. Note that this behavior is the same as it was before my PR, I did not change it. You will find the same comparison here: https://github.com/zephyrproject-rtos/zephyr/pull/54030/files#diff-927fe840a66cec50c33e92952dbef313db723ed7a43bdea4603da8d3c312373dL646

desowin replied:
This is strange and somewhat questionable. Transfers less than max packet size generally end transfer and allowing application to read less than max packet size of received data is somewhat weird. However, because it was there already I'm fine with it remaining there for the time being (especially that this PR does fix real issues).

tmon-nordic commented:
> Let me know if that is a problem, I can either remove the const from this change, or add the content of PR #54193 into this PR.

Adding the commit to this PR seems appropriate.

purdeaandrei commented:
> > Let me know if that is a problem, I can either remove the const from this change, or add the content of PR #54193 into this PR.
> 
> Adding the commit to this PR seems appropriate.

It already has 2 approvals, so I think I'll leave it in its own PR, it looks like it will likely be merged much faster.

jfischer-no commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -625,26 +651,66 @@ int usb_dc_ep_read_wait(uint8_t ep, uint8_t *data,
+		if (setup->wLength != 0U) {
+			/* In the case of a control transfer, we want to prime the OUT endpoint
+			 * exactly once, to either:
+			 * 1) in the to_device case, to receive the data (only if wLength is not 0)
+			 * 2) in the to_host case, to receive a 0-length status stage transfer
+			 *    (only valid if wLength is not 0)
+			 * Note that when wLength = 0, the status stage transfer is always an IN
+			 * type so we don't need to consider that case.
+			 */
+			arm_out_endpoint = true;
+			state.control_transfer_bytes_transferred = 0;
+		}
+		state.setup_available = false;
+	} else {
+		const size_t bytes_received = udc_rpi_get_ep_buffer_len(ep);
+
+		LOG_DBG("EP %d received %d offset: %d", ep, bytes_received, ep_state->read_offset);
+		if (ep == USB_CONTROL_EP_OUT) {
+			const struct usb_setup_packet *setup =
+				(const void *)&usb_dpram->setup_packet;
+			if (usb_reqtype_is_to_device(setup)) {
+				if (state.control_transfer_bytes_transferred + ep_state->read_offset
+					< setup->wLength) {
+					/* If no more data in the buffer, but we're still waiting
+					 * for more, start a new read transaction.
+					 */
+					if (bytes_received == ep_state->read_offset) {
+						state.control_transfer_bytes_transferred +=
+							ep_state->read_offset;
+						arm_out_endpoint = true;
+					}
+				}
+			}
+		} else {
+			/* If no more data in the buffer, start a new read transaction. */
+			if (bytes_received == ep_state->read_offset) {
+				arm_out_endpoint = true;
+			}
+		}
+	}
```
Most of the new code is specific to control OUT endpoint, please move it to
`static int usb_dc_control_ep_read_continue(void)` function.

purdeaandrei replied:
DONE, I used the same name as you suggested, but I had to add some arguments, so it's not just (void)

purdeaandrei commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -56,6 +56,7 @@ struct udc_rpi_state {
 	struct udc_rpi_ep_state in_ep_state[USB_NUM_BIDIR_ENDPOINTS];
 	bool setup_available;
 	bool should_set_address;
+	uint16_t control_transfer_bytes_transferred;
```
DONE, it's now called control_out_ep_rcvd

purdeaandrei commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -120,6 +125,18 @@ static int udc_rpi_start_xfer(uint8_t ep, const void *data, size_t len)
 static void udc_rpi_handle_setup(void)
 {
 	struct cb_msg msg;
+	struct udc_rpi_ep_state *ep_state = udc_rpi_get_ep_state(USB_CONTROL_EP_OUT);
```
DONE

purdeaandrei commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -120,6 +125,18 @@ static int udc_rpi_start_xfer(uint8_t ep, const void *data, size_t len)
+		while (!(usb_hw->abort_done & (1 << 1)))
+			;
+		*ep_state->buf_ctl &= ~USB_BUF_CTRL_AVAIL;
+		hw_clear_alias(usb_hw)->abort = 1 << 1;
```
BIT(1) is the bit associated with EP0_OUT. I have factored this out a function udc_rpi_cancel_endpoint(), and there's no magic number BIT(1), instead now which bit(s) to set/clear/check are determined by a call to the new function udc_rpi_endpoint_mask(ep). Hope it's now clearer to understand.

purdeaandrei commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -120,6 +125,18 @@ static int udc_rpi_start_xfer(uint8_t ep, const void *data, size_t len)
+	struct udc_rpi_ep_state *ep_state = udc_rpi_get_ep_state(USB_CONTROL_EP_OUT);
+
+	if (*ep_state->buf_ctl & USB_BUF_CTRL_AVAIL) {
+		/* Un-arm EP0_OUT endpoint, to make sure next setup packet starts clean */
```
DONE. Note that as I was writing the comment I realized that the more detailed comment didn't really make sense without also including a change that I was planning for a following PR, so I have included that here as well. I have also updated the commit message with the following lines added to the end:
     - When a setup req interrupt fires, don't immediately force the next
       read to get it, instead, it will be read only after a setup message
       is extracted from the message queue.
     - When a setup packet is received abort any unfinished previous control
       transfers:
        - cancel any data buffers given to the EP0_IN endpoint
        - drop any new ep0_in writes that are attempted before this newest
          setup packet's associated message is extracted from the message
          queue.

purdeaandrei replied:
So please see the new beginning of the udc_rpi_handle_setup() call and see if it makes sense.

purdeaandrei commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -419,6 +436,15 @@ int usb_dc_ep_set_stall(const uint8_t ep)
+			;
+		*udc_rpi_get_ep_state(USB_CONTROL_EP_OUT)->buf_ctl &= ~USB_BUF_CTRL_AVAIL;
+		hw_clear_alias(usb_hw)->abort = 1 << 1;
+	}
```
DONE, please see udc_rpi_cancel_endpoint()

purdeaandrei commented:
Note that I have also added the following fixes for race conditions that are tightly related, i.e. they can happen during control transfers.

These new race conditions only showed up in my testing when the system was heavily slowed down by blocking uart logging.

- When a setup req interrupt fires, don't immediately force the next
       read to get it, instead, it will be read only after a setup message
       is extracted from the message queue.
- In the ISR, check buffer interrupts before setup req interrupts.
       This is to make sure that the final 0-length status message from the
       previous setup packet is consumed before the new setup packet.
       (this is the only case now when both interrupts could be seen as
       having fired by the time the interrupt handler routine executes.

petejohanson commented:
Don't have the bandwidth to offer a meaningful review of the code changes right now, but got some quick testing in tonight, and using these changes, along w/ the few other PRs from @purdeaandrei and this is rock stable on ZMK, with both HID and CDC ACM device set up w/ the logging system and ouputing a ton on both endpoints.

jfischer-no commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -625,26 +733,85 @@ int usb_dc_ep_read_wait(uint8_t ep, uint8_t *data,
+			return ret;
+		}
+	} else {
+		const size_t bytes_received = udc_rpi_get_ep_buffer_len(ep);
```
`		const size_t len = udc_rpi_get_ep_buffer_len(ep);
`

purdeaandrei replied:
DONE

jfischer-no commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -625,26 +733,85 @@ int usb_dc_ep_read_wait(uint8_t ep, uint8_t *data,
 
-	state.setup_available = false;
+		LOG_DBG("EP%d_OUT received %d offset: %d", USB_EP_GET_IDX(ep),
+			bytes_received, ep_state->read_offset);
```
```c
		LOG_DBG("Endpoint 0x%02x received %u offset: %u",
			ep, len, ep_state->read_offset);
```

purdeaandrei replied:
DONE

jfischer-no commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -625,26 +733,85 @@ int usb_dc_ep_read_wait(uint8_t ep, uint8_t *data,
-	LOG_DBG("received %d offset: %d", bytes_received, ep_state->read_offset);
-	if (bytes_received == ep_state->read_offset) {
+	if (arm_out_endpoint) {
+		LOG_DBG("Arming EP%d_OUT", USB_EP_GET_IDX(ep));
```
`LOG_DBG("Arming endpoint 0x%02x", ep);`

purdeaandrei replied:
DONE

jfischer-no commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -625,26 +733,85 @@ int usb_dc_ep_read_wait(uint8_t ep, uint8_t *data,
+		LOG_DBG("Arming EP%d_OUT", USB_EP_GET_IDX(ep));
 		return usb_dc_ep_start_read(ep, DATA_BUFFER_SIZE);
+	} else {
+		LOG_DBG("Not arming EP%d_OUT", USB_EP_GET_IDX(ep));
```
`LOG_DBG("Not arming endpoint 0x%02x", ep);`

purdeaandrei replied:
DONE

jfischer-no commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -625,26 +733,85 @@ int usb_dc_ep_read_wait(uint8_t ep, uint8_t *data,
+
+		LOG_DBG("EP0_OUT received %d offset: %d", bytes_received, ep_state->read_offset);
+		if (usb_reqtype_is_to_device(setup)) {
+			if (state.control_out_ep_rcvd + ep_state->read_offset < setup->wLength) {
+				/* If no more data in the buffer, but we're still waiting
+				 * for more, start a new read transaction.
+				 */
+				if (bytes_received == ep_state->read_offset) {
+					state.control_out_ep_rcvd += ep_state->read_offset;
+					*arm_out_endpoint = true;
+				}
+			}
+		}
+	}
```
It looks like new functionality or where does it come from? The stack can handle it transaction by transaction.

purdeaandrei replied:
- The outer if(), is indeed new functionality. The very first version of this commit only re-armed the endpoint once after the setup packet is received. It was okay for short to_device transactions, however when wLength is longer than the maximum packet size, the out endpoint will need to be re-armed multiple times, until the whole data stage is consumed. The necessity of this was not exposed with my initial testing using keyboard firmware and set locklight messages, however I noticed this happening while testing the RNDIS app.

  When I pushed the second version of this commit,  I added this functionality and I have updated the commit message to say "for EP0 it is only re-armed if more OUT transactions are expected for that SETUP transfer (be it data-stage or status-stage)". In my opinion this line covers why line 770 is there but if you would like me to add more detail to the commit message, or comments, let me know.

- The inner if() is existing functionality, it matches line 646 of the original file: https://github.com/zephyrproject-rtos/zephyr/pull/54030/files#diff-927fe840a66cec50c33e92952dbef313db723ed7a43bdea4603da8d3c312373dL646
About that if() I already had a short conversation with desowin, see here: https://github.com/zephyrproject-rtos/zephyr/pull/54030#discussion_r1090263428
I do not feel comfortable to remove functionality as part of this PR even if I believed it was unused.

purdeaandrei commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -117,11 +123,85 @@ static int udc_rpi_start_xfer(uint8_t ep, const void *data, size_t len)
+
+	if (*ep_state->buf_ctl & USB_BUF_CTRL_AVAIL) {
+		udc_rpi_cancel_endpoint(USB_CONTROL_EP_OUT);
+
+		/* We are canceling EP0_OUT for the same reason as above, however this
+		 * warning could also be triggered by a software bug, if tight control
+		 * is not kept over when the ep0_out endpoint is armed. This endpoint must
+		 * only be armed when data packets are expected to be received, otherwise
+		 * we are exposed to race conditions when interrupt handling is a little
+		 * slow, and it becomes impossible to determine in what order did
+		 * setup/data transactions arrive.
+		 * Having the EP0_OUT endpoint be armed too early not only introduces
+		 * this hazard a single time, however it becomes sticky, because EP0_OUT
+		 * will be re-armed when the setup transaction is read, and because the
+		 * reading of the setup transaction happens in a thead, it will usually
+		 * happen after the host has sent its first try for a data transaction,
+		 * so the fresh re-arm will not be consumed by the current control
+		 * transfer, and ep0_out will perpetually remain armed.
+		 */
+
```
Do you mean to shorten the explanation? I can try, but I'm not sure if I can do it without removing information.

purdeaandrei replied:
DONE, and I ended up removing some information. Is it satisfacatory like this?

purdeaandrei commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -91,6 +93,10 @@ static int udc_rpi_start_xfer(uint8_t ep, const void *data, size_t len)
 	uint32_t val = len | USB_BUF_CTRL_AVAIL;
 
+	if (*ep_state->buf_ctl & USB_BUF_CTRL_AVAIL) {
+		LOG_WRN("ep %d was already armed", ep);
```
DONE

purdeaandrei commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -117,11 +123,85 @@ static int udc_rpi_start_xfer(uint8_t ep, const void *data, size_t len)
+ * @param [in] ep        Endpoint address
+ *
+ * @return a 32-bit mask specific to rpi_pico hardware
+ */
```
DONE

purdeaandrei commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -117,11 +123,85 @@ static int udc_rpi_start_xfer(uint8_t ep, const void *data, size_t len)
+ * @brief Cancel an already armed endpoint
+ *
+ * @param [in] ep        Endpoint address
+ */
```
DONE

purdeaandrei commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -117,11 +123,85 @@ static int udc_rpi_start_xfer(uint8_t ep, const void *data, size_t len)
+ *
+ * This function converts a zephyr endpoint address into a
+ * bit mask that can be used with registers:
+ *  - buf_status
```
DONE, I used the exact register name as it appears in modules/hal/rpi_pico/src/rp2040/hardware_structs/include/hardware/structs/usb.h
which does spell the register names differently from the datasheet, however, I am okay with using the datasheet names here, so I have made the change.

purdeaandrei commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -117,11 +123,85 @@ static int udc_rpi_start_xfer(uint8_t ep, const void *data, size_t len)
 	struct cb_msg msg;
 
-	state.setup_available = true;
+	/* Normally all control transfers should complete before a new setup transaction is
+	 * sent, however in some rare cases from the perspective of the device, a new setup
+	 * transaction could arrive prematurely, in which case the previous control transfer
+	 * should be aborted. See section 5.5.5 of the Universal Serial Bus Specification,
+	 * version 2.0.
```
DONE, I think. I have shortened the comments, however could you clarify what the exact rule is? Should indentation be included in those 80 characters? If yes, then how many characters should tab be treated as? 1/2/4/8? Should comments be more restricted in width than code below and above them? I was assuming before that my comment width was okay, because code around it was wider, and the PR checker didn't complain about any of it. It would be helpful if all of these rules were checked by the PR checker, then we wouldn't have to waste time on communication about this simple issue.

purdeaandrei commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -117,11 +123,85 @@ static int udc_rpi_start_xfer(uint8_t ep, const void *data, size_t len)
+	if (*ep_state->buf_ctl & USB_BUF_CTRL_AVAIL) {
+		udc_rpi_cancel_endpoint(USB_CONTROL_EP_OUT);
+
+		/* We are canceling EP0_OUT for the same reason as above, however this
```
DONE  (and after done the whole comment has been shortened)

purdeaandrei commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -117,11 +123,85 @@ static int udc_rpi_start_xfer(uint8_t ep, const void *data, size_t len)
+
+		/* We are canceling EP0_OUT for the same reason as above, however this
+		 * warning could also be triggered by a software bug, if tight control
+		 * is not kept over when the ep0_out endpoint is armed. This endpoint must
```
DONE  (and after done the whole comment has been shortened)

purdeaandrei commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -117,11 +123,85 @@ static int udc_rpi_start_xfer(uint8_t ep, const void *data, size_t len)
+		 * we are exposed to race conditions when interrupt handling is a little
+		 * slow, and it becomes impossible to determine in what order did
+		 * setup/data transactions arrive.
+		 * Having the EP0_OUT endpoint be armed too early not only introduces
```
DONE  (and after done the whole comment has been shortened)

purdeaandrei commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -117,11 +123,85 @@ static int udc_rpi_start_xfer(uint8_t ep, const void *data, size_t len)
+		 * slow, and it becomes impossible to determine in what order did
+		 * setup/data transactions arrive.
+		 * Having the EP0_OUT endpoint be armed too early not only introduces
+		 * this hazard a single time, however it becomes sticky, because EP0_OUT
```
DONE  (and after done the whole comment has been shortened)

purdeaandrei commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -625,26 +733,85 @@ int usb_dc_ep_read_wait(uint8_t ep, uint8_t *data,
+			*arm_out_endpoint = true;
+			state.control_out_ep_rcvd = 0;
+		}
+		state.setup_available = false;
```
DONE

purdeaandrei commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -625,26 +733,85 @@ int usb_dc_ep_read_wait(uint8_t ep, uint8_t *data,
+	} else {
+		const size_t bytes_received = udc_rpi_get_ep_buffer_len(USB_CONTROL_EP_OUT);
+
+		LOG_DBG("EP0_OUT received %d offset: %d", bytes_received, ep_state->read_offset);
```
DONE

purdeaandrei commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -595,8 +699,12 @@ int usb_dc_ep_read_wait(uint8_t ep, uint8_t *data,
+	if (ep == USB_CONTROL_EP_OUT && state.setup_available) {
 		read_count = sizeof(struct usb_setup_packet);
+		if (read_count != max_data_len) {
+			LOG_ERR("Attempting to read setup packet with the wrong length"
```
DONE, I changed it to a warning, I see this more as an assert, that would trigger, if there are any more read-ordering bugs in the driver.

purdeaandrei commented:
I added two more improvements to this PR:
1) I added a `k_sem_give()` call to `udc_rpi_cancel_endpoint()` to make sure any in endpoints that are canceled don't end up locked up.
2) I realized that there is USB Errata item RP2040-E2 that prevents using the abort registers on chip revisions B1/B0. I added conditional code to not use abort on these chips.

yonsch commented:
@jfischer-no can you take another look?

jfischer-no commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -625,26 +729,87 @@ int usb_dc_ep_read_wait(uint8_t ep, uint8_t *data,
 }
 
-int usb_dc_ep_read_continue(uint8_t ep)
+/*
+ * @brief utility function, grouping control ep code, called by usb_dc_ep_read_continue
+ *
+ * @param [in] ep_state          ep0_out state
+ * @param [out] arm_out_endpoint set to true, if more data transactions are expected
+ *
+ * @return error code
+ */
```
```suggestion
```

purdeaandrei commented:
@jfischer-no I committed your suggested change, and then squashed and rebased to fresh main.

jfischer-no commented:
@purdeaandrei Please remove Co-authored-by.
